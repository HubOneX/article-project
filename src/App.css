.App {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  min-height: 100vh;
  background: linear-gradient(
    333deg,
    rgba(36, 36, 36, 1) 0%,
    rgba(60, 60, 60, 1) 100%
  );
  color: #e6e6e6;
}
