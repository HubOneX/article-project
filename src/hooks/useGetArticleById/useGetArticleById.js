import axios from "axios";
import { useEffect, useState } from "react";
import { urls } from "../../constants/urls";

export const useGetArticleById = (id) => {
  const [article, setArticle] = useState({});
  const [isError, setIsError] = useState(false);

  const fetchArticleById = async (id) => {
    let data;

    await axios
      .get(urls.articleByIdUrl + id)
      .then((response) => {
        console.log({ response }, response.data.elements);
        const { elements } = response.data;
        data = {
          author: elements?.author?.value || "",
          heading: elements?.heading?.value || "",
          date: elements?.date?.value || "",
          image: {
            url: elements?.mainImage?.value?.leadImage?.renditions?.lead?.source
              ? urls.globalUrl +
                elements.mainImage.value.leadImage.renditions.lead.source
              : "",
            caption: elements?.mainImage?.value?.leadImageCaption?.value || "",
          },
          body: elements?.body?.values || "",
        };
      })
      .catch((error) => {
        console.error(error);
        setIsError(true);
      });

    if (data) {
      setArticle(data);
    }
  };

  useEffect(() => {
    fetchArticleById(id);
  }, [id]);

  return { article, isError };
};
