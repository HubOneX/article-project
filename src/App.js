import "./App.css";
import Article from "./components/article/Article";
import Loader from "./components/loader/Loader";
import { useGetArticleById } from "./hooks/useGetArticleById/useGetArticleById";
import { useEffect, useState } from "react";
import objectUtils from "./utils/objectUtils";

const givenId = "db4930e9-7504-4d9d-ae6c-33facca754d1";

function App() {
  const { article, isError } = useGetArticleById(givenId);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    if (article && objectUtils.isEmptyObject(article) && !isError) {
      setIsLoading(true);
      return;
    }

    if (isError || article) {
      setIsLoading(false);
    }
  }, [article, isError]);

  return (
    <div className="App">
      {isLoading ? <Loader /> : <Article article={article} isError={isError} />}
    </div>
  );
}

export default App;
