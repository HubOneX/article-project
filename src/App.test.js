import { act, render, screen, waitFor } from "@testing-library/react";
import App from "./App";
import * as fetchHook from "./hooks/useGetArticleById/useGetArticleById";
import {
  mockArticle,
  mockArticleEmptyValues,
  mockArticleResponse,
} from "./mocks/mockArticle";
import axios from "axios";
import renderer from "react-test-renderer";

// mocking axios to clean up console output from errors, a bit redundant
jest.mock("axios");
const mockedAxios = axios;

describe("App", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockedAxios.get.mockResolvedValueOnce(mockArticleResponse);
  });

  describe("Loader", () => {
    it("should be displayed on initial render", () => {
      render(<App />);
      const loader = screen.getByText("Loading...");
      expect(loader).toBeInTheDocument();
    });

    it("should be displayed when article is an empty object and there isError = false", async () => {
      jest.spyOn(fetchHook, "useGetArticleById").mockImplementation(() => {
        return {
          article: {},
          isError: false,
        };
      });

      render(<App />);
      await waitFor(() => {
        const loader = screen.queryByText("Loading...");
        expect(loader).toBeInTheDocument();
      });
    });

    it("should not be displayed when article is present", async () => {
      jest.spyOn(fetchHook, "useGetArticleById").mockImplementation(() => {
        return {
          article: mockArticle,
        };
      });

      render(<App />);
      await waitFor(() => {
        const loader = screen.queryByText("Loading...");
        expect(loader).not.toBeInTheDocument();
      });
    });

    it("should not be displayed when isError is TRUE", () => {
      jest.spyOn(fetchHook, "useGetArticleById").mockImplementation(() => {
        return {
          isError: true,
        };
      });

      render(<App />);
      const loader = screen.queryByText("Loading...");
      expect(loader).not.toBeInTheDocument();
    });
  });

  it("should match the snapshot when article has all values", () => {
    jest.spyOn(fetchHook, "useGetArticleById").mockImplementation(() => {
      return {
        article: mockArticle,
      };
    });

    let app;

    act(() => {
      app = renderer.create(<App />);
    });

    expect(app).toMatchSnapshot();
  });

  it("should match the snapshot when article has only empty strings", () => {
    jest.spyOn(fetchHook, "useGetArticleById").mockImplementation(() => {
      return {
        article: mockArticleEmptyValues,
      };
    });

    let app;

    act(() => {
      app = renderer.create(<App />);
    });

    expect(app).toMatchSnapshot();
  });
});
