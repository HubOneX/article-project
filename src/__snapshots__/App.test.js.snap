// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot when article has all values 1`] = `
<div
  className="App"
>
  <div
    className="article-wrapper"
  >
    <img
      alt="main image"
      className="article-image"
      src="https://content-eu-4.content-cms.com/api/859f2008-a40a-4b92-afd0-24bb44d10124/url"
    />
    <div
      className="article-header"
    >
      <div
        className="article-info"
      >
        <p
          className="article-author"
        >
          Stephen King
        </p>
        <p
          className="article-date"
        >
          07:27 Sep 3, 2020
        </p>
      </div>
      <hr />
      <h1
        className="article-title"
      >
        The horror of tests
      </h1>
    </div>
    <div
      className="article-body"
    >
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<h2>Secondary heading</h2>",
          }
        }
      />
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<p>The paragraph below!</p>",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`App should match the snapshot when article has only empty strings 1`] = `
<div
  className="App"
>
  <div
    className="article-wrapper"
  >
    <div
      className="article-header"
    >
      <div
        className="article-info"
      >
        <p
          className="article-author"
        >
          Anonymous
        </p>
        <p
          className="article-date"
        >
          -
        </p>
      </div>
      <hr />
      <h1
        className="article-title"
      />
    </div>
    <p
      className="not-loaded-info"
    >
      Sorry, we couldn't load this article. Please try again later.
    </p>
  </div>
</div>
`;
